src/main/scala/io/iohk/ethereum/domain/Blockchain.scala

```diff
@@ -323,20 +358,64 @@ class BlockchainImpl(
       )
     }
 
+    val (checkpointUpdates, prevCheckpointNumber): (DataSourceBatchUpdate, Option[BigInt]) = maybeBlockHeader match {
+      case Some(header) =>
+        if (header.hasCheckpoint && header.number == getLatestCheckpointBlockNumber()) {
+          val prev = findPreviousCheckpointBlockNumber(header.number, header.number)
+          prev.map { num =>
+            (appStateStorage.putLatestCheckpointBlockNumber(num), Some(num))
+          }.getOrElse {
+            (appStateStorage.removeLatestCheckpointBlockNumber(), Some(0))
+          }
+        } else (appStateStorage.emptyBatchUpdate, None)
+      case None =>
+        (appStateStorage.emptyBatchUpdate, None)
+    }
+
+    val newBestBlockNumber: BigInt = if(bestSavedBlock >= 1) bestSavedBlock - 1 else 0
+
     blockHeadersStorage.remove(blockHash)
       .and(blockBodiesStorage.remove(blockHash))
       .and(totalDifficultyStorage.remove(blockHash))
       .and(receiptStorage.remove(blockHash))
       .and(maybeTxList.fold(transactionMappingStorage.emptyBatchUpdate)(removeTxsLocations))
       .and(blockNumberMappingUpdates)
+      .and(appStateStorage.putBestBlockNumber(newBestBlockNumber))
+      .and(checkpointUpdates)
       .commit()
 
     // not transactional part
+    saveBestKnownBlocks(newBestBlockNumber, prevCheckpointNumber)
```

Comment:
Should we also add this to our save function as well?

Reply:
in `save` is different logic which controls what blocks data will be saved, so it's better to use low-level methods there

Reply:
And much more error prone... I think we'll suffer from this eventually :confused: